Add Documentation Re: Tool Functionalization and Enhance Behavior

qskwood opened the pull request:
These commits add some documentation including letting customers know they can install optional tools by setting an environment variable `SKIP_TOOLS=0` and making it clear these will be skipped if they're not useful on the detected target machine. There is also some internal documentation regarding both old and new functionality.

The script now checks whether an executable (e.g., `sensors`) is available and stores that information in the tools array as metadata. This is done whether the tool installation is expected or not, to allow for future access to that information.[^1] Metadata about executable check status is stored in the tools array using a new update_needed_tools function which can be used to more easily and safely update this or other metadata in the psuedo-multidimensional array.

The script also does some sanity checking on the array to reduce the likelihood of someone entering data into it incorrectly (it currently checks for two locations that should always be integers) and warns the user to report to support in the event of a failed check. The script does similar sanity checking on the new function to update the tools array data; if fields must be integers the script contains a function that can be used to cleanly refuse to proceed when invalid updates to the array are requested.

[^1]: For example, I intend to use that information to enhance output files when executables don't exist. Right now the output files may end up containing a warning but only if they are empty when a check is run. This metadata should be a better way to communicate that warning to support.

qskwood committed:
fix: add missing variable containing package name to install notification

The script was missing a variable containing the package name when notifying the user a package
was being installed. This adds the package name to the output.

qskwood committed:
feat: add argument numbers to tool array

The tool array did not previously reference the argument numbers that are used later. This adds the
argument numbers in the usage documentation.

qskwood committed:
feat: create executable check

Previously the script suppressed output when a tool was not available, without storing information
about missing tools anywhere. This adds metadata for the tool executable check (pass/fail) which can
be accessed later on along with a function to update the tool metadata easily. Logic for the install
checks has been adjusted to ensure this happens regardless of whether tool installation is needed.

qskwood committed:
feat: add integer check and apply to update_needed_tools

Previously the script did not perform any santity checking to ensure two provided values were
integers. Now the script will exit if they are not and will instruct the user to report the issue to
Lambda support for further review. These errors would happen if the function were called incorrectly
by someone who edited the script and should not occur because of anything during runtime.

qskwood committed:
fix: remove return codes

I did not intend to use these return codes, so I am removing them.

qskwood committed:
refactor: modify function name to be more descriptive

This fuction does not necessarily install anything, it first checks whether tools are present and
whether they should be installed. The new function name better reflects the behavior.

qskwood committed:
feat: validate the tools array makes sense

Previously the script did no checking and invalid information could be provided in the tools array.
The script now checks two of the metadata fields to ensure they are integers and exits with a
warning in the event they are not. This should only happen if the array is updated with invalid
information and should not otherwise occur during runtime.

qskwood committed:
docs: update script info output regarding optional tool installation

The script did not previously tell users how to allow tool installation, now it does. The script
also now makes it clear these will be skipped if unlikely to be useful on the target machine and
outputs the relevant package list in alphabetical order (instead of array order.)

bryangwin commented:
I think it should be more of a here is what we recommend to install... would you like to install these? Telling the customer after they already run the script that they need to run it again by setting an environment variable if they want tools installed is a poor UX.

qskwood commented:
I can add a prompt to this shortly.

qskwood commented:
I took a look at this but I don't think that's a feature we can trivially add unless I'm mistaken. If we want customers to use `curl` to grab the script and pipe it to Bash, the IO redirection will break our ability to use something like the `read` builtin.

Two options:
* leave this as-is, preserving the non-interactive nature of the script (curl, pipe, bash)
  * customers need to select the option proactively or run it a second time
  * could updating the docs site with this information to mitigate the above
* modify the current workflow to allow both non-interactive and interactive selection of the option
  * download, then run
  * docs site would need updated
  * customers would potentially keep the downloaded script, meaning they end up running an outdated copy later

I would prefer to keep this as-is.

bryangwin commented:
@qskwood 

If they are running in an interactive shell, which I think most users are, I believe you should be able do this:
```
echo "The following tools are missing and could be installed: ${TOOLS[*]}"

# Check if we can open /dev/tty (interactive shell?)
if head -c 0 /dev/tty 2>/dev/null; then
    echo -n "Would you like to install them now? [y/N]: "
    read -r answer < /dev/tty
    if [[ "${answer}" =~ ^[Yy]$ ]]; then
        export SKIP_TOOLS=0
    fi
else
    echo ""
    echo "No interactive shell detected."
    echo "The script will continue running without installing the missing tools."
    echo "If you'd like to install them, re-run the script with SKIP_TOOLS=0 set in your environment."
    echo ""
    echo "Example:"
    echo "SKIP_TOOLS=0 wget -nv -O - https://raw.githubusercontent.com/lambdal-support/lambda-public-tools/main/lambda-bug-report.sh | bash -"
    echo ""
fi
```

Note: Not tested

qskwood committed:
feat: provide interactive prompt regarding tool installation

Previously if a user ran the script without allowing tool installation, they would need to run it a
second time to allow the install. Now the script gives the user an opportunity to allow them while
running, unless already overridden.

qskwood commented:
Perfect! I plugged this into my attempt and have committed the update to the branch.

bryangwin commented:
Awesome. @jashoakley can you test this out and approve if it all looks good?

jashoakley reviewed (approved):
Tested the following:
- Ran without installing tools option
- Ran with installing tools option
- Ran with SKIP_TOOLS=0
